### DISCUSSION TITLE
converting Reports

### DISCUSSION THREAD

**Discussion by glevkovich** [Ericsson/codechecker#4078] [General]
Hi,

I’m trying to convert the the output of Google Sanitizers into CodeChecker support format (plist). Instruction are posted here: 
https://codechecker.readthedocs.io/en/latest/tools/report-converter/#install-guide

I’m using Codechecker 6.22 (installed with pip), and tried converting with both gcc-11 and Clang multiple Clang versions.
I've followed all the instructions, including the commands:
```
# Create a Python virtualenv and set it as your environment.
make venv
source $PWD/venv/bin/activate

# Build and install report-converter package.
make package

```

I've done that by  cloning the codechecker repository and checkout the same tag installed, in order to  be able to execute "make package".
I always get errors of the type:
```
(CodeChecker venv) root@62a4509d2672:/# report-converter -t tsan -o ./tsan_results ./tsan.output
[INFO] - No 'tsan' results can be found in '/tsan.output'.
[WARNING] - Use '--meta' option to provide extra information to the CodeChecker server such as analyzer version and analysis command when storing the results to it. For more information see the --help.
```

Here is an example for a TSAN output file which I can convert with the simple above command, can someone please help me and tell me what i'm doing wrong? It just doesn't work..

```
(CodeChecker venv) root@62a4509d2672:/# cat tsan.output
==================
WARNING: ThreadSanitizer: data race (pid=1941)
  Read of size 4 at 0x55ef65e421d4 by thread T2:
    #0 increment() //a.cpp:6 (a.out+0x1295)
    #1 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) /usr/include/c++/12/bits/invoke.h:61 (a.out+0x1f3e)
    #2 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) /usr/include/c++/12/bits/invoke.h:96 (a.out+0x1ea3)
    #3 void std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/include/c++/12/bits/std_thread.h:252 (a.out+0x1e08)
    #4 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() /usr/include/c++/12/bits/std_thread.h:259 (a.out+0x1db2)
    #5 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() /usr/include/c++/12/bits/std_thread.h:210 (a.out+0x1d6c)
    #6 <null> <null> (libstdc++.so.6+0xd44a2)

  Previous write of size 4 at 0x55ef65e421d4 by thread T1:
    #0 increment() //a.cpp:6 (a.out+0x12ad)
    #1 void std::__invoke_impl<void, void (*)()>(std::__invoke_other, void (*&&)()) /usr/include/c++/12/bits/invoke.h:61 (a.out+0x1f3e)
    #2 std::__invoke_result<void (*)()>::type std::__invoke<void (*)()>(void (*&&)()) /usr/include/c++/12/bits/invoke.h:96 (a.out+0x1ea3)
    #3 void std::thread::_Invoker<std::tuple<void (*)()> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/include/c++/12/bits/std_thread.h:252 (a.out+0x1e08)
    #4 std::thread::_Invoker<std::tuple<void (*)()> >::operator()() /usr/include/c++/12/bits/std_thread.h:259 (a.out+0x1db2)
    #5 std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (*)()> > >::_M_run() /usr/include/c++/12/bits/std_thread.h:210 (a.out+0x1d6c)
    #6 <null> <null> (libstdc++.so.6+0xd44a2)

  Location is global 'shared_data' of size 4 at 0x55ef65e421d4 (a.out+0x41d4)

  Thread T2 (tid=1944, running) created by main thread at:
    #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1001 (libtsan.so.2+0x5e686)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd4578)
    #2 main //a.cpp:11 (a.out+0x130d)

  Thread T1 (tid=1943, running) created by main thread at:
    #0 pthread_create ../../../../src/libsanitizer/tsan/tsan_interceptors_posix.cpp:1001 (libtsan.so.2+0x5e686)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd4578)
    #2 main //a.cpp:10 (a.out+0x12f7)

SUMMARY: ThreadSanitizer: data race //a.cpp:6 in increment()
==================
Shared data: 2000000
ThreadSanitizer: reported 1 warnings
```

**Comment by glevkovich** [comment]
An update: I did the "build package" on the codechecker repository origin/master, and while in the Python virtualenv tried the above command again, with the same results.

**Comment by bruntib** [comment]
The report-converter tool uses the following regex for finding thread sanitizer output:
```.py
    # Regex for parsing ThreadSanitizer output message.
    line_re = re.compile(
        # Error code
        r'==(?P<code>\d+)==(ERROR|WARNING): ThreadSanitizer: '
        # Checker message.
        r'(?P<message>[\S \t]+)')
```
I'm not sure if the attached thread sanitizer producing an output that doesn't match this regex is because of a different sanitizer version, or different configuration. Could you try adding the suggested flags for gcc when running sanitizer? (`-fno-omit-frame-pointer`) https://github.com/Ericsson/codechecker/blob/master/docs/tools/report-converter.md#sanitizers

If the regex still doesn't match, then report-converter should be extended with matching this new format.

**Comment by harsht1737** [comment]
Hi @bruntib 

Tried executing report-converter with output from compilation with given flags and still got the same error as @glevkovich posted above.

```
root@010157feb63c:/#
root@010157feb63c:/# g++ -fsanitize=thread -g -fno-omit-frame-pointer data_race.cpp -std=c++11 -o data_race
root@010157feb63c:/#
root@010157feb63c:/# ./data_race > 2tsan.output 2>&1
root@010157feb63c:/#
```

```
(CodeChecker venv) root@010157feb63c:/codechecker/build/CodeChecker/bin#
(CodeChecker venv) root@010157feb63c:/codechecker/build/CodeChecker/bin# ./report-converter -t tsan -o /codechecker/tsan_results /codechecker/2tsan.output
[INFO] - No 'tsan' results can be found in '/codechecker/2tsan.output'.
[WARNING] - Use '--meta' option to provide extra information to the CodeChecker server such as analyzer version and analysis command when storing the results to it. For more information see the --help.
(CodeChecker venv) root@010157feb63c:/codechecker/build/CodeChecker/bin#
(CodeChecker venv) root@010157feb63c:/codechecker/build/CodeChecker/bin#
```

Also, tried with sample asan log file provided under report-converter's test directory and that also seems to provide same error.

```

(CodeChecker venv) root@010157feb63c:/codechecker/build/CodeChecker/bin#
(CodeChecker venv) root@010157feb63c:/codechecker/build/CodeChecker/bin# ./report-converter -t asan -o /codechecker/asan_results /codechecker/tools/report-converter/tests/unit/analyzers/asan_output_test_files/asan.out
[INFO] - No 'asan' results can be found in '/codechecker/tools/report-converter/tests/unit/analyzers/asan_output_test_files/asan.out'.
[WARNING] - Use '--meta' option to provide extra information to the CodeChecker server such as analyzer version and analysis command when storing the results to it. For more information see the --help.
(CodeChecker venv) root@010157feb63c:/codechecker/build/CodeChecker/bin#
```

Can you please let us know if anything is incorrect/missing in these steps.
Or, Could you please share correct set of steps to convert a tsan log output to plist using report-converter tool for this cpp file -data_race.cpp (attached txt)

[data_race.txt](https://github.com/Ericsson/codechecker/files/13311625/data_race.txt)

**Comment by glevkovich** [comment]
Hi @bruntib,

I don't think it has anything with a "new format".
Here is Tsan official format page, there are a few formats supported:
https://github.com/google/sanitizers/wiki/ThreadSanitizerReportFormat
`"Format of the reports differ depending on bug type (see [ThreadSanitizerDetectableBugs](https://github.com/google/sanitizers/wiki/ThreadSanitizerDetectableBugs))."`

Furthermore, your python pytest unit test is passing, but if the input files from the tests for both ASAN/TSAN are taken and used externally by report-converter, it fails with the same error. So I suspect that something might be wrong with another assertion/validation logic which is not being run in the unit test, but is part of the the report-converter (probably a bug).

I am not sure about other sanitizers such as MSan UBsan, and LSan, we might check them too.
